### ISSUE TITLE
Use Ppx_deriving_show directly

### ISSUE THREAD

**Issue by tekknolagi** [whitequark/ppx_deriving#122]
I would like to use the `Ppx_deriving_show` module directly, for the `str_of_type` and `sig_of_type` functions. I don't see any documentation on how to link it properly. What is the best way to do that?

**Comment by whitequark** [comment]
You can't, the side effect of using this module is registration of a deriver. What's the use case?

**Comment by tekknolagi** [comment]
I am building a programming language that essentially has ppx_deriving.show built-in. It compiles to OCaml. In the parsing stage it's just another one of several AST transforms. It would be nice to get access to those two functions.

**Comment by tekknolagi** [comment]
Right now we vendor ppx_deriving_show.ml directly, which is an ugly hack.

**Comment by whitequark** [comment]
It seems like vendoring is a better option here. Consider: say I add a new option to ppx_deriving_show. Now the public interface of your language changes, without you having done anything. Is this good?

**Comment by tekknolagi** [comment]
Currently we pin version 4.1, so as far as I understand, the interface of the language would not change.

**Comment by whitequark** [comment]
So does that mean using the reason package and any other version of ppx_deriving together is impossible? That seems annoying, except if Reason always lives in its own opam switch.

**Comment by tekknolagi** [comment]
I don't think so, but I very well could be wrong. I am not super familiar with how the OCaml ecosystem works just yet. But here is why I think that could be OK. The use for Reason is at the moment something like:

1. Write Reason code
2. Use the provided `refmt` (which bundles in ppx_deriving.show) to transform the code into OCaml code (or JS, I suppose), optionally with generated printer functions

Since it's just `refmt` that uses the transform and not any other part of the toolchain, it seems to me as though other versions of ppx_deriving could be used at other points in time. Please correct me if I am wrong.

If we would rather not vendor, what would you say is the best way to do an AST transform manually? Or is the only option to vendor?

**Comment by whitequark** [comment]
There can only be one version of ppx_deriving (or any other package) installed in opam at any one time. So if you choose to depend on ppx_deriving, there are two options:
  * Either the user is forced to remain on the version of ppx_deriving that Reason uses (and I'm quite uncomfortable with that, because it effectively means ppx_deriving is now frozen);
  * Or whichever version of ppx_deriving is used in Reason implicitly becomes a part of its public interface.

In other words:
   * Either I lose control,
   * Or you lose control.

Which part of the Reason toolchain performs the conversion is immaterial.

> If we would rather not vendor, what would you say is the best way to do an AST transform manually?

You could be a "ppx driver" and add the `[@@deriving show]` annotation before running ppx_deriving.

**Comment by tekknolagi** [comment]
@whitequark We've vendored ppx_deriving and ppx_deriving_show. Cheers!
